src/tcp.rs

```diff
@@ -463,47 +481,49 @@ impl Connection {
                 self.send.una = ackn;
             }
 
-            // TODO: prune self.unacked
             // TODO: if unacked empty and waiting flush, notify
             // TODO: update window
         }
 
         if let State::FinWait1 = self.state {
-            if self.send.una == self.send.iss + 2 {
-                // our FIN has been ACKed!
-                self.state = State::FinWait2;
+            if let Some(closed_at) = self.closed_at {
+                if self.send.una == closed_at.wrapping_add(1) {
+                    // our FIN has been ACKed!
+                    self.state = State::FinWait2;
+                }
             }
         }
 
-        if let State::Estab | State::FinWait1 | State::FinWait2 = self.state {
-            let mut unread_data_at = (self.recv.nxt - seqn) as usize;
-            if unread_data_at > data.len() {
-                // we must have received a re-transmitted FIN that we have already seen
-                // nxt points to beyond the fin, but the fin is not in data!
-                assert_eq!(unread_data_at, data.len() + 1);
-                unread_data_at = 0;
+        if !data.is_empty() {
+            if let State::Estab | State::FinWait1 | State::FinWait2 = self.state {
+                let mut unread_data_at = self.recv.nxt.wrapping_sub(seqn) as usize;
+                if unread_data_at > data.len() {
+                    // we must have received a re-transmitted FIN that we have already seen
+                    // nxt points to beyond the fin, but the fin is not in data!
+                    assert_eq!(unread_data_at, data.len() + 1);
+                    unread_data_at = 0;
+                }
+                self.incoming.extend(&data[unread_data_at..]);
+
+                /*
+                Once the TCP takes responsibility for the data it advances
+                RCV.NXT over the data accepted, and adjusts RCV.WND as
+                apporopriate to the current buffer availability.  The total of
+                RCV.NXT and RCV.WND should not be reduced.
+                 */
+                self.recv.nxt = seqn.wrapping_add(data.len() as u32);
```

Comment:
Why does this not need to take into account the FIN bit if we set it?

Reply:
Specifically, given that we _overwrite_ `self.recv.nxt` here, we do not add to it, so the `+1` you added further down may be overwritten (I think).

Reply:
First, motivation for changes to this part of code - they were not strictly necessary (nothing crashed), but I added them anyway to get rid of annoying duplicate ACKs (specifically, previous version acked the empty ACK of the handshake and acked their FIN twice).

The new code more closely (I think) follows the RFC - first process the segment data (p.73), advance recv.nxt, send appropriate ACK, then process the FIN if it is set (p.74), advance recv.nxt, send appropriate ACK. It might send ACKs twice but they would be different ACKs (one for data, one for FIN), so technically it is ok I think.

> Why does this not need to take into account the FIN bit if we set it?

Not sure I understood this correctly - do you mean *our* or *their* FIN? Their FIN is processed further down (only in FIN_WAIT2 because everything else is unsupported :)) I think there might be some problems if the segment arrived out of order (and there is a gap between recv.nxt and seqn) but I don't think we handle this case well anyway.

> Specifically, given that we overwrite self.recv.nxt here, we do not add to it, so the +1 you added further down may be overwritten (I think).

I didn't test specifically with packets from the other side that contain both data and FIN but I think it should be ok? FIrst recv.nxt is advanced here to the end of the received data (and an ACK is sent). And then if there is also FIN, recv.nxt is advanced once more at the stage where the FIN is processed.

Reply:
Ah, I was thinking about _their_ FIN. Specifically, you now update `self.recv.nxt` here: https://github.com/jonhoo/rust-tcp/pull/3/files#diff-9432f854ab6fc747c10c8e50dc46aa14R526, but reading through the code again, I don't think there's actually an issue since we already check whether the ACK is "acceptable" before we handle it.

Thanks for the detailed reply!